### PR TITLE
Issue #1888: added in the \dfrac macro for displaystyle fractions

### PR DESCRIPTION
**Pull request by watkinrt**
This is a partial fix for issue #1888. The implementation of \dfrac{}{} addresses the issue of creating fractions with displaystyle formatting (which the author of #1888 would like to be able to do); however, it does not implement the more generic \displaystyle functionality as explicitly requested in #1888 (I've begun coding this up, but it is reasonably involved and has a number of pitfalls). 

I added in an example script (dfrac_demo.py) to demonstrate the new functionality.

### CODE REVIEW

**Comment by dstansby** [comment]
There's a couple of PEP8 violations in the test - some trailing whitespace at the end of the lines, and no newline at the end of the file (see https://travis-ci.org/matplotlib/matplotlib/jobs/205418695#L1677)

File: examples/text_labels_and_annotations/dfrac_demo.py [hunk 1]
```diff
@@ -0,0 +1,15 @@
+#!/usr/bin/env python2
```
Comment (tacaswell) [inline examples/text_labels_and_annotations/dfrac_demo.py]: remove the pinning to python2

**Comment by tacaswell** [comment]
Can you please add a test to https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/tests/test_mathtext.py#L17  to cover this case?

Over all :+1: thanks for working on this!

File: examples/text_labels_and_annotations/dfrac_demo.py (hunk 1 above)
Comment (watkinrt) [inline examples/text_labels_and_annotations/dfrac_demo.py]: Will do. This is my first contribution, so I'm still working out some of the bugs. Thanks for the prompt suggestions.

**Comment by tacaswell** [comment]
To get the test images you will have to add the test, run the test suite, let if it fail on a missing image, then copy output into `tests/baseline_images/test_mathtext` and commit them (see http://matplotlib.org/devel/testing.html#writing-an-image-comparison-test)

You will also need to be sure to be testing against the correct version of freetype (see http://matplotlib.org/devel/contributing.html#retrieving-and-installing-the-latest-version-of-the-code)

File: examples/text_labels_and_annotations/dfrac_demo.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+# -*- coding: utf-8 -*-
+"""
+Created on Sat Feb 25 17:50:37 2017
```
Comment (QuLogic) [inline examples/text_labels_and_annotations/dfrac_demo.py]: Remove this date; we have `git log` for that. Can you please modify the docstring to follow the [sphinx-gallery format](http://matplotlib.org/devdocs/devel/MEP/MEP12.html#clean-up-guidelines)?

File: examples/text_labels_and_annotations/dfrac_demo.py [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+@author: watkinrt
+"""
+
+import matplotlib.pyplot as plt
+
+fig = plt.figure(figsize=(5.25, 0.75))
+fig.text(0.5, 0.3, r'\dfrac: $\dfrac{a}{b}$',
+         horizontalalignment='center', verticalalignment='center')
+fig.text(0.5, 0.7, r'\frac: $\frac{a}{b}$',
+         horizontalalignment='center', verticalalignment='center')
```
Comment (QuLogic) [inline examples/text_labels_and_annotations/dfrac_demo.py]: Needs `plt.show` at the end.

File: lib/matplotlib/mathtext.py [hunk 4]
```diff
@@ -3077,35 +3087,46 @@ def _genfrac(self, ldelim, rdelim, rule, style, num, den):
         return result
 
     def genfrac(self, s, loc, toks):
-        assert(len(toks)==1)
-        assert(len(toks[0])==6)
+        assert(len(toks) == 1)
+        assert(len(toks[0]) == 6)
```
Comment (QuLogic) [inline lib/matplotlib/mathtext.py]: 👍 for the whitespace fix, but also, you don't really need the parentheses.

File: examples/text_labels_and_annotations/dfrac_demo.py (hunk 3 above)
Comment (watkinrt) [inline examples/text_labels_and_annotations/dfrac_demo.py]: Man, you guys are fast and thorough. I'm working on all of the requested changes.

File: lib/matplotlib/mathtext.py [hunk 5]
```diff
@@ -3043,8 +3050,11 @@ def _genfrac(self, ldelim, rdelim, rule, style, num, den):
             state.font, state.fontsize, state.dpi)
 
         rule = float(rule)
-        num.shrink()
-        den.shrink()
+
+        # If style != displaystyle == 0, shrink the num and den
```
Comment (QuLogic) [inline lib/matplotlib/mathtext.py]: I think these warrant some constants instead of magic numbers.

**Comment by tacaswell** [comment]
''power-cycled' to trigger travis.

**Comment by QuLogic** [comment]
It needs to be a raw string _or_ escaped backslashes, not both.

**Comment by watkinrt** [comment]
Am I missing something? The reStructuredText documents make it sound like you need to make the text is a raw string to properly pass the backslashes out of python, but that you still need to escape the backslashs for the reStructuredText to be parsed correctly (http://docutils.sourceforge.net/docs/user/rst/quickref.html#escaping).

**Comment by QuLogic** [comment]
At this point, we don't actually render any of that text through Sphinx. After #8247 is merged, then maybe it will be clearer what should be done.

File: examples/text_labels_and_annotations/dfrac_demo.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
+r"""
+=========================================
+The difference between \dfrac and \frac
+=========================================
+
+In this example, the differences between the \dfrac and \frac TeX macros are
+illustrated; in particular, the difference between display style and text style
+fractions when using Mathtex. aasdsads
```
Comment (anntzer) [inline examples/text_labels_and_annotations/dfrac_demo.py]: mathtext

**Comment by QuLogic** [comment]
Do you want to have a go at rebasing this? The sphinx-gallery stuff is merged now, so we should be able to see whether the escaping is going to be a problem.

**Comment by watkinrt** [comment]
I should be able to get to it this weekend.

**Comment by tacaswell** [comment]
It looks like some of the new required test images did not get added 

```
_________________ test_mathtext_rendering[png-mathtext-cm-82] __________________
[gw1] linux -- Python 3.4.6 /home/travis/build/matplotlib/matplotlib/venv/bin/python
self = <matplotlib.testing.decorators._ImageComparisonBase object at 0x7fffe12af160>
baseline = 'mathtext_cm_82', extension = 'png'
    def copy_baseline(self, baseline, extension):
        baseline_path = os.path.join(self.baseline_dir, baseline)
        orig_expected_fname = baseline_path + '.' + extension
        if extension == 'eps' and not os.path.exists(orig_expected_fname):
            orig_expected_fname = baseline_path + '.pdf'
        expected_fname = make_test_filename(os.path.join(
            self.result_dir, os.path.basename(orig_expected_fname)), 'expected')
        if os.path.exists(orig_expected_fname):
            shutil.copyfile(orig_expected_fname, expected_fname)
        else:
            reason = ("Do not have baseline image {0} because this "
                      "file does not exist: {1}".format(expected_fname,
                                                        orig_expected_fname))
>           raise ImageComparisonFailure(reason)
E           matplotlib.testing.exceptions.ImageComparisonFailure: Do not have baseline image /home/travis/build/matplotlib/matplotlib/result_images/test_mathtext/mathtext_cm_82-expected.png because this file does not exist: /home/travis/build/matplotlib/matplotlib/lib/matplotlib/tests/baseline_images/test_mathtext/mathtext_cm_82.png
lib/matplotlib/testing/decorators.py:271: ImageComparisonFailure
```

I suspect the issue is missing optional dependencies on the local system where the test images were generated.  @watkinrt I can try to generate them and push to this branch or you can sort out which deps you are missing (my guess is LaTeX?), which ever you would prefer.

**Comment by watkinrt** [comment]
> @watkinrt I can try to generate them and push to this branch or you can sort out which deps you are missing (my guess is LaTeX?), which ever you would prefer.
I just realized I hadn't fetched the master branch from the main matplotlib repository in a while; I image that is why I was missing some of the new test images. I've updated my working copy and just pushed it. I haven't been able to successfully run the full test suite on this commit as something is messed up with the my development copy of matplotlib. If someone else has time, would you mind running the tests? I'm not sure when I'll be able to get around to fixing my development copy.

Sorry for all of my newbiew mistakes; this is the first large scale repository/team I've worked on.

**Comment by QuLogic** [comment]
It's preferable to rebase instead of merging `master`. However, the tests are failing because it looks like you forgot to commit the image for the new test.

**Comment by tacaswell** [comment]
@watkinrt I'll try to generate those test images for you now and push to your branch.

With your permission I can do the rebase for you (and force-push to your branch).

Don't worry about small process mistakes, we were all new once!

**Comment by tacaswell** [comment]
I pushed a commit and then almost immediately force-pushed a fixed version of it (accidentally committed a temporary file).  Hopefully no one fetched between the two pushes, but if you did, that is the source of any issue.

**Comment by tacaswell** [comment]
Looks like this is passing now and the test images look good!

**Comment by watkinrt** [comment]
Great! Thanks for closing all of this out. Hopefully that wraps everything up for this pull request.

**Comment by tacaswell** [comment]
@QuLogic I think this is ready to go, can you give this another review or dismiss your change request?

**Comment by tacaswell** [comment]
@watkinrt Thanks and congratulations on what (I think) is your first Matplotlib contribution :tada: !

Thanks for taking on one of the more arcane bits of the code base and bearing with us through a very long review process! 

Hopefully we will hear from you again.

**Comment by breedlun** [comment]
Wow, this is great, @watkinrt !  I had no idea you had finished this almost 2 years ago!

**Pull request merged**